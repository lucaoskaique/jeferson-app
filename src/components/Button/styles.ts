import styled, { css, DefaultTheme } from 'styled-components'

import { ButtonProps } from '.'

type WrapperProps = {
  hasIcon: boolean
} & Pick<ButtonProps, 'size' | 'background' | 'fontWeight'>

const wrapperModifiers = {
  xlarge: (theme: DefaultTheme) => css`
    height: 6rem;
    font-size: ${theme.font.sizes.xlarge};
    padding: ${theme.spacings.xsmall} ${theme.spacings.xxlarge};
  `,
  large: (theme: DefaultTheme) => css`
    height: 5rem;
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xlarge};
  `,
  medium: (theme: DefaultTheme) => css`
    height: 4rem;
    font-size: ${theme.font.sizes.small};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.medium};
  `,
  small: (theme: DefaultTheme) => css`
    height: 3rem;
    font-size: ${theme.font.sizes.xsmall};
  `,
  ice: () => css`
    background: linear-gradient(90deg, #72c042 -99.98%, #0092d3 193.73%);
  `,
  hot: () => css`
    background: linear-gradient(90deg, #ffbf3a -40%, #d90726 185.31%);
  `,
  fontWeight: (theme: DefaultTheme) => css`
    font-weight: ${theme.font.bold};
  `,
  withIcon: (theme: DefaultTheme) => css`
    svg {
      width: 3rem;
      & + span {
        margin-left: ${theme.spacings.xxsmall};
      }
    }
  `
}

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, size, background, fontWeight, hasIcon }) => css`
    display: inline-flex;
    align-items: center;
    justify-content: center;
    color: ${theme.colors.white};
    border: 0;
    cursor: pointer;
    font-family: ${theme.font.family};
    border-radius: ${theme.border.radius};
    padding: ${theme.spacings.xxsmall};
    text-decoration: none;

    ${!!size && wrapperModifiers[size](theme)}
    ${!!background && wrapperModifiers[background]}
    ${!!fontWeight && wrapperModifiers.fontWeight(theme)}
    ${!!hasIcon && wrapperModifiers.withIcon(theme)}
  `}
`
