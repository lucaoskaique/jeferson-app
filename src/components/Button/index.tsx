import { forwardRef, AnchorHTMLAttributes, ButtonHTMLAttributes } from 'react'

import * as S from './styles'

type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>

export type ButtonProps = {
  background?: 'hot' | 'ice'
  size?: 'xlarge' | 'large' | 'medium' | 'small'
  icon?: React.ReactNode
  fullWidth?: boolean
  fontWeight?: boolean
  as?: React.ElementType
} & ButtonTypes

const Button: React.ForwardRefRenderFunction<S.WrapperProps, ButtonProps> = (
  {
    children,
    size = 'large',
    background = 'hot',
    fullWidth = false,
    icon,
    ...props
  },
  ref
) => (
  <S.Wrapper
    size={size}
    background={background}
    fullWidth={fullWidth}
    hasIcon={!!icon}
    ref={ref}
    {...props}
  >
    {!!icon && icon}
    {!!children && <span>{children}</span>}
  </S.Wrapper>
)

export default forwardRef(Button)
