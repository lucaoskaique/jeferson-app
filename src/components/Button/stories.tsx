import { Story, Meta } from '@storybook/react'
import { DownloadForOffline } from '@styled-icons/material-rounded/DownloadForOffline'

import Button, { ButtonProps } from '.'

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    children: {
      type: 'string'
    }
  }
} as Meta

export const Default: Story<ButtonProps> = (args) => <Button {...args} />

Default.args = {
  children: 'Botão'
}

export const withIcon: Story<ButtonProps> = (args) => <Button {...args} />

withIcon.args = {
  size: 'small',
  children: 'Botão',
  icon: <DownloadForOffline />
}
