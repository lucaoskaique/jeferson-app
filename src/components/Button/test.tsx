import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/tests/helpers'

import Button from '.'

describe('<Button />', () => {
  it('should render hot gradient background Button by default', () => {
    renderWithTheme(<Button>Deputado Jeferson Fernandes</Button>)
    expect(
      screen.getByRole('button', { name: /deputado Jeferson Fernandes/i })
    ).toHaveStyle({
      background: 'linear-gradient(90deg, #ffbf3a -40%, #d90726 185.31%)'
    })
  })

  it('should render ice gradient background Button', () => {
    renderWithTheme(
      <Button background="ice">Deputado Jeferson Fernandes</Button>
    )
    expect(
      screen.getByRole('button', { name: /deputado Jeferson Fernandes/i })
    ).toHaveStyle({
      background: 'linear-gradient(90deg, #72c042 -99.98%, #0092d3 193.73%)'
    })
  })

  it('should render a xlarge default Button', () => {
    renderWithTheme(<Button size="xlarge">Button</Button>)
    expect(screen.getByRole('button', { name: /Button/i })).toHaveStyle({
      height: '7rem',
      padding: '1.6rem 5.6rem',
      'font-size': '2.0rem'
    })
  })

  it('should render a large default Button', () => {
    renderWithTheme(<Button>Button</Button>)
    expect(screen.getByRole('button', { name: /Button/i })).toHaveStyle({
      height: '5rem',
      padding: '0.8rem 4.8rem',
      'font-size': '1.6rem'
    })
  })

  it('should render a medium Button', () => {
    renderWithTheme(<Button size="medium">Button</Button>)
    expect(screen.getByRole('button', { name: /Button/i })).toHaveStyle({
      height: '4rem',
      padding: '0.8rem 3.2rem',
      'font-size': '1.4rem'
    })
  })

  it('should render a small Button', () => {
    renderWithTheme(<Button size="small">Button</Button>)
    expect(screen.getByRole('button', { name: /Button/i })).toHaveStyle({
      height: '3rem',
      'font-size': '1.2rem'
    })
  })

  it('should render a Button with font-weight bold', () => {
    renderWithTheme(<Button fontWeight>Button</Button>)
    expect(screen.getByRole('button', { name: /Button/i })).toHaveStyle({
      'font-weight': '600'
    })
  })
})
