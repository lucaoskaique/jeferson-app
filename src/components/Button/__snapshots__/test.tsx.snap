// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[
  `<Button /> should render hot gradient background Button by default 1`
] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  color: #FAFAFA;
  border: 0;
  cursor: pointer;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  border-radius: 0.4rem;
  padding: 0.8rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 600;
  height: 5rem;
  font-size: 1.6rem;
  padding: 0.8rem 4.8rem;
  background: linear-gradient(90deg,#ffbf3a -40%,#d90726 185.31%);
}

.c0:hover {
  background: linear-gradient(90deg,#72c042 -99.98%,#0092d3 193.73%);
}

<button
  class="c0"
>
  <span>
    Deputado Jeferson Fernandes
  </span>
</button>
`
